### PR TITLE
Allow ORT backend for DTensor

### LINKED ISSUE #101911: DTensor support for ORT
### 🚀 The feature, motivation and pitch

`DTensor` currently only supports cuda and cpu. The ort backend can be supported by removing some asserts and hardcoded values. 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ashari4**
fixes #101911

Currently, `DTensor` supports cuda and cpu. This PR makes some changes for easier integration with the ort backend.

* `Backend.NAME`  attribute now has value `name` instead of `NAME` for backends registered through `register_backend(name)`; this matches the pattern for backends with built-in support like nccl. 
* remove unused `_check_for_nccl_backend` function 
* add test case that moves parameters to device in the `partition_fn` - a scenario that's useful for big models

### CODE REVIEW

File: torch/distributed/_tensor/api.py [hunk 1]
```diff
@@ -519,7 +516,6 @@ def replicate_module_params_buffers(m: nn.Module, mesh: DeviceMesh) -> None:
         # apply partition_fun to submodules
         for name, submod in module.named_modules():
             partition_fn(name, submod, device_mesh)
-            replicate_module_params_buffers(submod, device_mesh)
```
Comment (ashari4) [inline torch/distributed/_tensor/api.py:522]: revert this change

File: torch/distributed/_tensor/placement_types.py [hunk 2]
```diff
@@ -148,22 +148,24 @@ def _shard_tensor(
...
         scatter_list, pad_sizes = self._split_tensor(
             tensor, num_chunks, with_padding=True, contiguous=True
         )
+        src_shard = scatter_list[my_coordinate[mesh_dim]]
+        output = torch.empty_like(src_shard, device=mesh.device_type)
 
-        output = torch.empty_like(scatter_list[my_coordinate[mesh_dim]])
-        mesh.scatter(output, scatter_list, mesh_dim=mesh_dim)
+        if not tensor.is_meta:
+            output.copy_(src_shard)
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:168]: I don't think this is the correct way to do sharding, as we need to preserve a global tensor semantic, this means to also preserve the global tensor content. Otherwise if we simply just copy the corresponding slice on each rank, it would results in some correctness issue.

i.e. by default if you do `torch.randn` on each rank, it would result in different data as a result of different seed/rng of the rank, so each rank would produce a big tensor that purely based on their random seed. Given this, we can't simply copy corresponding slice as that would break the contract about a "global tensor" distribute to different devices.

File: torch/distributed/_tensor/placement_types.py [hunk 3]
```diff
@@ -302,21 +304,22 @@ def __str__(self) -> str:
...
+            return tensor.new_empty(
+                0, device=mesh.device_type, requires_grad=tensor.requires_grad
+            )
 
-        tensor = tensor.contiguous()
-        mesh.broadcast(tensor, mesh_dim=mesh_dim)
-        return tensor
+        output = torch.empty_like(tensor, device=mesh.device_type)
+        if not tensor.is_meta:
+            output.copy_(tensor)
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:320]: Similarly, we can't simply copy the corresponding replica as that would result in different rank have different data, even though the intention of `replicate_tensor` is to replicate the data.

Random seed is important to ensure model correctness so we would have to do a broadcast in this case.

File: torch/distributed/_tensor/device_mesh.py [hunk 4]
```diff
@@ -167,9 +166,11 @@ def _get_or_create_default_group(self):
             # as an attribute of device mesh for future use. However, the detail is still TBD how we gonna use
             # this attribute, so we will implement this logic once we figure out the answer.
             self._seed = torch.cuda.initial_seed()
+        elif self.device_type == "ort":
```
Comment (wanchaol) [inline torch/distributed/_tensor/device_mesh.py:169]: curious is `ort` both a device_type and also a communication backend? would `tensor.to("ort")` works?

File: torch/distributed/distributed_c10d.py [hunk 5]
```diff
@@ -572,6 +573,21 @@ class GroupMember(metaclass=_WorldMeta):
 
 STORE_BASED_BARRIER_PREFIX = "store_based_barrier_key"
 
+# DO NOT USE this OBJECT DIRECTLY.
+# Use the APIs below instead.
+_default_backend_for_device: Dict[str, Backend] = {
```
Comment (wanchaol) [inline torch/distributed/distributed_c10d.py:578]: @H-Huang what's the canonical way to register a backend for a device type in c10d?

**Review by wanchaol — CHANGES_REQUESTED** [review]
Thanks for contributing! Adding a new backend is exciting, but I think we should not change the scatter/broadcast logic due to the randomness issue, please see inline comments.

File: torch/distributed/_tensor/placement_types.py (hunk 2 above)
Comment (ashari4) [reply torch/distributed/_tensor/placement_types.py:168]: I see.

> i.e. by default if you do torch.randn on each rank, it would result in different data as a result of different seed/rng of the rank, so each rank would produce a big tensor that purely based on their random seed.

I see. yes, the approach of requiring the caller to call `torch.manual_seed` is a bit fragile right? also, it may not work for the multi-threaded case. I can revert it. 

But I noticed that every rank copies the entire tensor to the device:

https://github.com/pytorch/pytorch/blob/76af22103b4b886ca5d124d1df2e6cf8f53065f3/torch/distributed/_tensor/api.py#LL406C1-L407C52

 what do you think of doing this only in the source rank?

File: torch/distributed/_tensor/placement_types.py (hunk 3 above)
Comment (ashari4) [reply torch/distributed/_tensor/placement_types.py:320]: yes, thanks. i will revert it

File: torch/distributed/_tensor/device_mesh.py (hunk 4 above)
Comment (ashari4) [reply torch/distributed/_tensor/device_mesh.py:169]: `ort` is a device type so `tensor.to("ort")` will work. but `ort` is not a communication backend.

File: torch/distributed/_tensor/placement_types.py (hunk 2 above)
Comment (ashari4) [reply torch/distributed/_tensor/placement_types.py]: i've reverted the change to use scatter once again

Comment (ashari4) [reply torch/distributed/_tensor/placement_types.py]: based on our discussion, we will let every rank copy the entire tensor to preserve spmd-style programming.

File: torch/distributed/_tensor/device_mesh.py (hunk 4 above)
Comment (wanchaol) [reply torch/distributed/_tensor/device_mesh.py:164]: I think we don't need this special check anymore after https://github.com/pytorch/pytorch/pull/102336

File: torch/distributed/_tensor/placement_types.py [hunk 6]
```diff
@@ -150,20 +150,28 @@ def _shard_tensor(
...
+            )
 
         scatter_list, pad_sizes = self._split_tensor(
             tensor, num_chunks, with_padding=True, contiguous=True
         )
 
-        output = torch.empty_like(scatter_list[my_coordinate[mesh_dim]])
-        mesh.scatter(output, scatter_list, mesh_dim=mesh_dim)
+        output = torch.empty_like(
+            scatter_list[my_coordinate[mesh_dim]], device=mesh.device_type
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:171]: I think we don't need to pass device to this `empty_like`  call, `empty_like` would just reuse the same device with its input

File: torch/distributed/_tensor/placement_types.py [hunk 7]
```diff
@@ -150,20 +150,28 @@ def _shard_tensor(
...
+        )
+        if not tensor.is_meta:
+            mesh.scatter(output, scatter_list, mesh_dim=mesh_dim)
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:174]: inside `mesh.scatter` we already skip the call so we don't need this additional check?

File: torch/distributed/_tensor/placement_types.py [hunk 8]
```diff
@@ -308,15 +316,27 @@ def _replicate_tensor(
...
-            return tensor.new_empty(0, requires_grad=tensor.requires_grad)
+            return tensor.new_empty(
+                0, device=mesh.device_type, requires_grad=tensor.requires_grad
+            )
 
         tensor = tensor.contiguous()
-        mesh.broadcast(tensor, mesh_dim=mesh_dim)
-        return tensor
+        if tensor.is_meta:
+            return torch.empty_like(tensor, device=mesh.device_type)
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:332]: same here, I think we should only do `to_empty` inside `distribute_module` call, and don't eagerly initialize the memory inside `distribute_tensor` (i.e. _shard_tensor/_replicate_tensor)

**Review by wanchaol — COMMENTED** [review]
Have a few comments on the meta initialization, I think we should just do `to_empty` inside `distribute_module` without forcing the placement_types to eagerly initialize data if its local shards are on meta device. 

Also, based on our discussion, I am working on removing the DeviceMesh hard coded types and use the c10d dispatchable backend directly in https://github.com/pytorch/pytorch/pull/102336 So after this change we should get a easier backend integration.

File: torch/distributed/_tensor/device_mesh.py (hunk 4 above)
Comment (ashari4) [reply torch/distributed/_tensor/device_mesh.py]: thanks, I've removed it

File: torch/distributed/_tensor/placement_types.py (hunk 6 above)
Comment (ashari4) [reply torch/distributed/_tensor/placement_types.py]: I've removed it

File: torch/distributed/_tensor/placement_types.py (hunk 7 above)
Comment (ashari4) [reply torch/distributed/_tensor/placement_types.py]: i've removed the check

File: torch/distributed/_tensor/placement_types.py (hunk 8 above)
Comment (ashari4) [reply torch/distributed/_tensor/placement_types.py]: I've removed it

File: torch/distributed/distributed_c10d.py (hunk 5 above)
Comment (ashari4) [reply torch/distributed/distributed_c10d.py]: @wanchaol I've removed this change. but i think we need to sync on how `DeviceMesh` will create process groups for a non-nccl and non-gloo backend.

**Review by wanchaol — APPROVED** [review]
lgtm, thanks!
